cartographer_ros/cartographer_ros/metrics/internal/gauge.h

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2018 The Cartographer Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CARTOGRAPHER_ROS_METRICS_INTERNAL_GAUGE_H
+#define CARTOGRAPHER_ROS_METRICS_INTERNAL_GAUGE_H
+
+#include <atomic>
+#include <map>
+#include <string>
+
+#include "cartographer/metrics/gauge.h"
+#include "cartographer_ros_msgs/Metric.h"
+
+namespace cartographer_ros {
+namespace metrics {
+
+class Gauge : public ::cartographer::metrics::Gauge {
+ public:
+  explicit Gauge(const std::map<std::string, std::string>& labels)
+      : labels_(labels) {}
+
+  void Decrement(const double by_value) override { Add(-1. * by_value); }
+
+  void Decrement() override { Decrement(1.); }
+
+  void Increment(const double by_value) override { Add(by_value); }
+
+  void Increment() override { Increment(1.); }
+
+  void Set(double value) override {
+    double expected = value_.load();
+    while (!value_.compare_exchange_weak(expected, value)) {
+      ;
+    }
+  }
+  double Value() const { return value_.load(); }
+
+  cartographer_ros_msgs::Metric ToRosMessage() {
+    cartographer_ros_msgs::Metric msg;
+    msg.type = cartographer_ros_msgs::Metric::TYPE_GAUGE;
+    for (const auto& label : labels_) {
+      cartographer_ros_msgs::MetricLabel label_msg;
+      label_msg.key = label.first;
+      label_msg.value = label.second;
+      msg.labels.push_back(label_msg);
+    }
+    msg.value = Value();
+    return msg;
+  }
+
+ private:
+  void Add(const double value) {
+    double current = Value();
```

Comment:
@MichaelGrupp 
This is clearly a race condition: Two threads could read the current value, execute "current + value" in parallel, and then whoever calls Set first wins. The other increment is lost.

Reply:
Agreed, I think it would be better to

```
double expected = value_.load();
double new_value = expected + value;
 while (!value_.compare_exchange_weak(expected, new_value)) {
  new_value = expected + value;
}
```

Reply:
At a second glance at this code, I think we can even remove `Add` completely and directly use `std::atomic::fetch_add` and `std::atomic::fetch_sub` for increment and decrement, something like this:
```c++
  void Decrement(const double by_value) override { 
    value_.fetch_sub(by_value, std::memory_order_relaxed); 
  }

  void Decrement() override { Decrement(1.); }

  void Increment(const double by_value) override { 
    value_.fetch_add(by_value, std::memory_order_relaxed); 
  }

  void Increment() override { Increment(1.); }
```

Reply:
Do you agree?

Reply:
SG. I would rename by_value to e.g. decrement_value or something.

Reply:
Rollback: these operations are only supported for integral types....

Reply:
Is there a reason you want to do this with atomics instead of using the less error-prone solution using a mutex? It seems unlikely you need any potential erformance benefit here.

Reply:
@CodeArno FYI metrics are now fixed with https://github.com/googlecartographer/cartographer_ros/commit/fb5b16d6d3b496482221bd60785be106419fa79d